File: Adv Programming C & C++/practice_with_lists/Lex.java
import java.io.*;
import java.util.Scanner;

public class Lex{
	public static void main(String[] args){
		if(args.length==2){
			String inFile = args[0];
			String outFile = args[1];
			try{
				Scanner in = new Scanner(new File(inFile));
				PrintWriter out = new PrintWriter(new FileWriter(outFile));
				int i = 0;
				while(in.hasNextLine()){
					i++;
					in.nextLine();
				}
				String[] lines = new String[i];
				in = new Scanner(new File(inFile));
				i=0;
				while(in.hasNextLine()){
					lines[i] = in.nextLine();
					i++;
				}
				List list = new List();
				list.append(0);
				list.moveFront();
				/*for(int j=0;j<lines.length;j++){
					System.out.println(lines[j]);
				}*/
				for(int k=1; k<lines.length; k++){
					boolean high = true;
					//System.out.println("Placing "+lines[k]);
					list.moveFront();
					if(list.index()==(list.length()-1)){
						//System.out.println("Comparing "+lines[k]+" and "+lines[list.get()]);
						if(lines[k].compareTo(lines[list.get()])<0){
							high=false;
						}
					}
					while(high&&list.index()<(list.length()-1)){
						//System.out.println("Comparing "+lines[k]+" and "+lines[list.get()]);
						if(lines[k].compareTo(lines[list.get()])<=0){
							//System.out.println("Found place");
							high=false;
							break;
						}
						list.moveNext();
					}
					if(list.index()==(list.length()-1)){
						//System.out.println("Comparing "+lines[k]+" and "+lines[list.get()]);
						if(lines[k].compareTo(lines[list.get()])<0){
							high=false;
						}
					}
					if(high){
						list.append(k);
						//System.out.println("Inserted at the end");
					}else{
						list.insertBefore(k);
						//System.out.println("Inserted at position "+(list.index()-1));
					}
					
				}
				/*for(list.moveFront();list.index()>=0;list.moveNext()){
					System.out.println(list.get()+" "+lines[list.get()]);
				}*/
				for(list.moveFront();list.index()>=0;list.moveNext()){
					out.println(lines[list.get()]);
				}
				out.close();
			} catch(IOException e){
				System.err.println("Could not open files for read/write");
				System.exit(1);
			}	
		} else{
			System.err.println("Incorrect number of arguments");
		}




	}
}

File: Adv Programming C & C++/Lex/Lex.c
//-----------------------------------------------------------------------------
// FileIO.c
// Illustrates file input-output commands and text processing using the
// string functions strtok and strcat.
//-----------------------------------------------------------------------------

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<ctype.h>
#include "List.h"

#define MAX_LEN 160

typedef char string[MAX_LEN];


int main(int argc, char * argv[]){

   int count=0;
   FILE *in, *out;
   string line;
   string tokenlist;
   char* token;
   string lines[10240];

   // check command line for correct number of arguments
   if( argc != 3 ){
      //printf("Usage: %s <input file> <output file>\n", argv[0]);
      exit(1);
   }

   // open files for reading and writing 
   in = fopen(argv[1], "r");
   out = fopen(argv[2], "w");
   if( in==NULL ){
      //printf("Unable to open file %s for reading\n", argv[1]);
      exit(1);
   }
   if( out==NULL ){
      //printf("Unable to open file %s for writing\n", argv[2]);
      exit(1);
   }
   int i=0;
   while( fgets(line, MAX_LEN, in) != NULL)  {
      count++;
      token = strtok(line, " \n");
      tokenlist[0] = '\0';
      if(token!=NULL){
         strcat(tokenlist, token);
         token = strtok(NULL, " \n");
      }
      while( token!=NULL ){
         strcat(tokenlist, " ");
         strcat(tokenlist, token);
         //strcat(tokenlist, "\n");
         token = strtok(NULL, " \n");
      }
      strcpy(lines[i],tokenlist);
      //printf("Token: %s, tokenlist: %s, line:%s, count:%i\n",token,tokenlist,line,count);
      //fprintf(stderr, "tokenlist:%s  lines[%i]:%s count:%i  \n",tokenlist,i, lines[i], count);
      i++;
   }
   /*
   for(int j=0;j<count;j++){
      fprintf(stdout, "%s\n", lines[0]);
   }*/
   //printf("\n");
   
   List list = newList();
   append(list, 0);
   moveFront(list);

   for(int k=1; k<count; k++){
      int high =1;
      string str1;
      string str2;
      strcpy(str1, lines[k]);
      strcpy(str2, lines[get(list)]);
      //fprintf(stdout,"Finding spot for str1: %s\n", str1);
      moveFront(list);
      if(index(list)==(length(list)-1)){
         //printf("Comparing strings %s and %s\n", str1, str2);
         if(strcmp(str1,str2)<0){
            high=0;
         }
      }else{
         while(high==1&&index(list)<(length(list)-1)){
            strcpy(str2,lines[get(list)]);
            //printf("Comparing strings %s and %s\n", str1, str2);
            if(strcmp(str1,str2)<=0){
               //printf("Found place!  ");
               high=0;
               break;
            }
            moveNext(list);
         }
         if(index(list)==(length(list)-1)){
            strcpy(str2,lines[get(list)]);
            //printf("Comparing strings %s and %s\n", str1, str2);
            if(strcmp(str1,str2)<0){
               high=0;
            }
         }
      }
      
      if(high==1){
         append(list, k);
         //printf("Inserted at the end\n");
      }else{
         insertBefore(list, k);
         //printf("Inserted at position %i\n", (index(list)-1));
      }
      
   }
   
   for(moveFront(list);index(list)>=0;moveNext(list)){
      fprintf(out, "%s\n", lines[get(list)]);
   }
   
   freeList(&list);
   
   fclose(out);
   fclose(in);

   return(0);
}
